feat: Scenario events (Excluding exporting)

UberWaffe opened the pull request:
README for scenario events added. "res/manual/README_map_editor_scenario_events.txt" Add scenario events, which are a set of actions that can be triggered if a set of conditions are met. Conditions can be things like months passed, population greater than, etc. Actions can be things like, give or take money, unlock building, change trade price, etc.


Note: Removed xml exporting as requested, as it isn't up to snuff.

crudelios commented on src/core/xml_parser.h:
```diff
@@ -100,6 +100,20 @@ int xml_parser_get_attribute_bool(const char *key);
+ * 
+ * @return Name of the current element.
+ */
+const char *xml_parser_get_current_element_name();
```
```suggestion
const char *xml_parser_get_current_element_name(void);
```

UberWaffe replied:
Done.

crudelios commented on src/empire/city.c:
```diff
@@ -369,8 +381,20 @@ int empire_unlock_all_resources(void)
+    return 0;
+}
+
+int empire_city_our_city_change_resource_available(int resource, int new_state)
```
```suggestion
int empire_city_change_own_resource_availability(resource_type resource, int is_available)
```

UberWaffe replied:
Done.

crudelios reviewed (changes requested):
First review chunk. I looked at some non important files and have some small suggestions.

Expect some more reviews as I look through the changes.

UberWaffe commented on src/core/xml_parser.c:
```diff
@@ -352,6 +352,16 @@ int xml_parser_get_attribute_enum(const char *key,
+    return XML_GetCurrentLineNumber(data.parser);
+}
+
+const char *xml_parser_get_current_element_name()
```
Done.

UberWaffe commented:
Just as a sidenote.
Once this is in an acceptable state, I'll squash and rebase this on the latest master before merging.

crudelios commented on src/game/tick.c:
```diff
@@ -113,6 +114,8 @@ static void advance_month(void)
     city_gods_update_blessings();
     tutorial_on_month_tick();
+    scenario_events_progress_paused(1);
+    scenario_events_process_all();
```
I was wondering how often you would be checking for the scenario events. A monthly check seems perfect 👍

crudelios commented on src/scenario/action_types/action_handler.c:
```diff
@@ -0,0 +1,88 @@
+        case ACTION_TYPE_TRADE_SET_PRICE: return scenario_action_type_trade_price_set_execute(action);
+        default:
+            return 0;
+            break;
```
Nitpick: you don't really need the `break` after `return`ing.
Also, since the function names are rather large, I think they would look better on their own line instead of being on the `case` line.

crudelios commented on src/scenario/action_types/rome_wages.c:
```diff
@@ -0,0 +1,47 @@
+    city_data.labor.months_since_last_wage_change = 0;
+
+    int current_wages = city_labor_wages_rome();
+    if (is_hard_set
```
Same here. For small multiple conditionals we don't use a newline. We only use it if the conditions breaks over the 120 char line limit.

UberWaffe replied:
I'll remember that for the future and do a quick search to fix the current ones like this.
Done (with this one).

crudelios commented on src/scenario/action_types/send_standard_message.c:
```diff
@@ -0,0 +1,13 @@
+{
+    int32_t text_id = action->parameter1;
+
+    city_message_post(1, text_id, 0, 0);
```
You should check if the `text_id` is valid to prevent crashes. Remember: never trust user input.

crudelios commented on src/scenario/action_types/trade_price.c:
```diff
@@ -0,0 +1,24 @@
+    int32_t adjustment = action->parameter2;
+    int32_t show_message = action->parameter3;
+
+    int successfully_changed = trade_price_change(resource, adjustment);
```
You should also return here if `adjustment` is `0`.

UberWaffe replied:
Done.

crudelios commented on src/scenario/action_types/trade_price.c:
```diff
@@ -0,0 +1,24 @@
+
+    int successfully_changed = trade_price_change(resource, adjustment);
+    
+    if (successfully_changed
```
Single line, like above please.

crudelios commented on src/scenario/action_types/trade_price_set.c:
```diff
@@ -0,0 +1,41 @@
+
+    int successfully_changed = trade_price_change(resource, adjustment);
+    
+    if (successfully_changed
```
Single line here too.

UberWaffe replied:
Done.

crudelios commented on src/scenario/action_types/trade_route_amount.c:
```diff
@@ -0,0 +1,34 @@
+    int32_t show_message = action->parameter4;
+
+    trade_route_set_limit(route_id, resource, amount);
+    if (show_message
```
Single line here please.

UberWaffe replied:
Done.

crudelios commented on src/scenario/action_types/trade_route_amount.c:
```diff
@@ -0,0 +1,34 @@
+    int32_t resource = action->parameter2;
+    int32_t amount = action->parameter3;
+    int32_t show_message = action->parameter4;
+
```
You should check if the `route_id` or the `resource` are valid before executing this, otherwise this may end up crashing the game due to invalid data.

Maybe when there's a problem you should display a warning in the log.

UberWaffe replied:
Done.

UberWaffe commented on src/scenario/action_types/change_allowed_buildings.c:
```diff
@@ -0,0 +1,15 @@
+
+int scenario_action_type_change_allowed_buildings_execute(scenario_action_t *action)
+{
+    int32_t building = action->parameter1;
```
Done.

UberWaffe commented on src/game/file_io.c:
```diff
@@ -410,6 +428,22 @@ static void get_version_data(savegame_version_data *version_data, savegame_versi
     }
+
+    if (version > SAVE_GAME_LAST_NO_EVENTS) {
+        version_data->piece_sizes.scenario_events = PIECE_SIZE_DYNAMIC;
```
Done.

UberWaffe commented on src/scenario/action_types/change_allowed_buildings.c:
Done.

UberWaffe commented:
That should be everything from the current review batch.
I've moved actions into one file. Same for conditions.

crudelios reviewed (changes requested):
Another bunch of comments.

I'm getting to the core of the scenario events finally! About 2000 lines of code left to check...

UberWaffe commented on src/empire/trade_route.c:
```diff
@@ -35,6 +35,12 @@ int trade_route_count(void)
+int trade_route_is_valid(int route_id)
+{
+    route_resource *route = array_item(routes, route_id);
+    return (route != 0);
```
Done.

UberWaffe commented on src/scenario/condition_types/comparison_helper.c:
```diff
@@ -0,0 +1,21 @@
+{
+    switch (compare_type) {
+        case COMPARISON_TYPE_EQUAL:
+            return (value1 == value2);
```
Done.

UberWaffe commented on src/scenario/condition_types/condition_handler.c:
```diff
@@ -0,0 +1,79 @@
+}
+
+void scenario_condition_type_load_state(buffer *buf,
+    int version, scenario_condition_t *condition, int *link_type, int32_t *link_id)
```
Done.

UberWaffe commented on src/scenario/condition_types/condition_types.c:
```diff
@@ -0,0 +1,282 @@
+
+int scenario_condition_type_building_count_active_met(scenario_condition_t *condition)
+{
+    int type = condition->parameter1;
```
Done.

UberWaffe commented on src/scenario/condition_types/condition_types.c:
```diff
@@ -0,0 +1,282 @@
+    int total_active_count = 0;
+    switch(building_type) {
+        case BUILDING_MENU_FARMS:
+            total_active_count += building_count_active(BUILDING_WHEAT_FARM);
```
The condition is called "building_count_active" in the xml.
So I meant to check for active.
I haven't added other count types yet, as I'm not sure they would be needed.
I'll look into doing that if once people use it they request it.

crudelios replied:
If that's the case, then it's correct.

I thought you wanted to count all buildings, not just the active ones.

UberWaffe commented on src/scenario/condition_types/condition_types.c:
```diff
@@ -0,0 +1,282 @@
+    int value = condition->parameter2;
+    int in_city_only = condition->parameter3;
+
+    int soldier_count = city_military_total_soldiers();
+    if (in_city_only) {
+        soldier_count = city_military_total_soldiers_in_city();
+    }
```
Done.

UberWaffe commented on src/scenario/condition_types/condition_types.c:
```diff
@@ -0,0 +1,282 @@
+
+int scenario_condition_type_request_is_ongoing_met(scenario_condition_t *condition)
+{
+    int request_id = condition->parameter1;
```
I added a check to the `scenario_request_is_ongoing` function. Let me know if I should instead have it here.

UberWaffe commented on src/scenario/condition_types/condition_types.c:
```diff
@@ -0,0 +1,282 @@
+#include "scenario/request.h"
+#include "scenario/condition_types/comparison_helper.h"
+
+int scenario_condition_type_building_count_active_met(scenario_condition_t *condition)
```
Done.

UberWaffe commented on src/scenario/condition_types/condition_types.c:
```diff
@@ -0,0 +1,282 @@
+{
+    int type = condition->parameter1;
+    int value = condition->parameter2;
+    int building_type = condition->parameter3;
```
Done.

UberWaffe commented on src/scenario/request.c:
```diff
@@ -14,6 +14,8 @@
 #include "game/tutorial.h"
 #include "scenario/data.h"
 
+void scenario_request_make_visible_and_send_message(int id);
```
Done.

UberWaffe commented on src/scenario/request.c:
```diff
@@ -174,6 +181,50 @@ const scenario_request *scenario_request_get_visible(int index)
+    return 0;
+}
+
+int scenario_request_force_request_start(int id)
```
Done.
And added sanity check for id range.

UberWaffe commented on src/window/plain_message_dialog.c:
```diff
@@ -93,7 +98,20 @@ void window_plain_message_dialog_show(translation_key title, translation_key mes
+    }
+}
+
+void window_plain_message_dialog_show_with_two_extra(translation_key title, translation_key message, const uint8_t *extra, const uint8_t *extra2)
```
Done.

UberWaffe commented on src/window/editor/attributes.c:
```diff
@@ -42,9 +43,11 @@ static void button_win_criteria(int param1, int param2);
 static void change_climate(int param1, int param2);
 static void change_image(int forward, int param2);
 
+#define NUMBER_OF_BUTTONS 11
```
Done.
Did other UI windows with similar code as well.

UberWaffe commented on src/translation/english.c:
```diff
@@ -909,6 +909,27 @@ static translation_string all_strings[] = {
+    {TR_EDITOR_CHECK_LOG_MESSAGE, "Please check augustus-log.txt in your Augustus directory for details."},
+    {TR_EDITOR_UNABLE_TO_SAVE_EVENTS_TITLE, "Unable to save scenario events"},
+    {TR_EDITOR_IMPORT_LINE, "Line: "},
+    {TR_EDITOR_SCENARIO_EVENT_ID, "ID : "},
```
Done.

UberWaffe commented:
Latest batch of comments done. (Some exceptions, see question / feedback on comments.)

crudelios commented on src/scenario/scenario_events_xml.h:
```diff
@@ -0,0 +1,638 @@
+    xml_data_attribute_t xml_parm5;
+} scenario_condition_data_t;
+
+static scenario_condition_data_t scenario_condition_data[CONDITION_TYPE_MAX] = {
```
These `static` values shouldn't be on a header file, but rather on a C file.

When you need to access, you should return a pointer to their ID via a function on the C file.

Having them as `static` on a header will cause all its values to be duplicated in memory for each different C file that includes this header, which is not what you want.

UberWaffe replied:
Allright, this was a pretty big refactor.
I moved all these from scenario_events_xml to a new file scenario_events_parameter_data.c
(Not sure on the naming, suggestions welcome.)

crudelios replied:
I imagined it would be, but it was a needed one.

Kudos for adding it to its own file, I think it makes sense that way.

I'll check it out.

crudelios commented on src/scenario/scenario_events_xml.h:
```diff
@@ -0,0 +1,638 @@
+    [CONDITION_TYPE_TIME_PASSED]     = { .type = CONDITION_TYPE_TIME_PASSED,
+                                        .xml_attr =     { .name = "time",           .type = PARAMETER_TYPE_TEXT },
+                                        .xml_parm1 =    { .name = "check",          .type = PARAMETER_TYPE_CHECK,            .min_limit = 1,         .max_limit = 3 },
+                                        .xml_parm2 =    { .name = "min",            .type = PARAMETER_TYPE_MIN_MAX_NUMBER,           .min_limit = 0,         .max_limit = 1000000000 },
```
Why not define an `UNLIMITED` that is set to `1000000000` instead'?

But just like I mentioned above, that define should be on the C file, not on the header.

UberWaffe replied:
Done.

crudelios commented on src/scenario/scenario_events_xml.h:
```diff
@@ -0,0 +1,638 @@
+#include "scenario/scenario_event_data.h"
+
+typedef enum {
+    PARAMETER_TYPE_UNDEFINED = -1,
```
I'd set `PARAMETER_TYPE_UNDEFINED` to `0` and increase from there.

It will be easier to handle uninitialized values that way.

UberWaffe replied:
Done.

crudelios commented on src/scenario/scenario_events_xml.h:
```diff
@@ -0,0 +1,638 @@
+                                        .xml_parm1 =    { .name = "check",          .type = PARAMETER_TYPE_CHECK,            .min_limit = 1,         .max_limit = 3 },
+                                        .xml_parm2 =    { .name = "min",            .type = PARAMETER_TYPE_MIN_MAX_NUMBER,           .min_limit = 0,         .max_limit = 1000000000 },
+                                        .xml_parm3 =    { .name = "max",            .type = PARAMETER_TYPE_MIN_MAX_NUMBER,           .min_limit = 0,         .max_limit = 1000000000 },
+                                        .xml_parm4 =    { .name = "",               .type = PARAMETER_TYPE_UNDEFINED },
```
If you set `PARAMETER_TYPE_UNDEFINED` to `0` on the `enum` above, you won't need to fill values for unused elements.

All you may have to check is for the `.xml_paramX` against null pointers instead of empty strings, but depending on your code you probably won't even need to change anything at all (I still haven't checked `scenario_events_xml.c` at the time of writing this).

UberWaffe replied:
I've done the undefined as 0.
Still looking into what the impact of not defaulting to empty string is.

UberWaffe replied:
Done. I've made the name change as well.

crudelios commented on src/scenario/scenario_events_xml.h:
```diff
@@ -0,0 +1,638 @@
+    int value;
+} special_attribute_mapping_t;
+
+#define SPECIAL_ATTRIBUTE_MAPPINGS_CHECK_SIZE 3
```
You can use the good old `sizeof(array name) / sizeof(array item)` trick here for the array count.

UberWaffe replied:
Done.

crudelios commented on src/scenario/scenario_events_xml.h:
```diff
@@ -0,0 +1,638 @@
+
+#define SPECIAL_ATTRIBUTE_MAPPINGS_CHECK_SIZE 3
+static special_attribute_mapping_t special_attribute_mappings_check[SPECIAL_ATTRIBUTE_MAPPINGS_CHECK_SIZE] = {
+    [0]     = { .type = PARAMETER_TYPE_CHECK,                .text = "eq",                            .value = COMPARISON_TYPE_EQUAL},
```
```suggestion
    { .type = PARAMETER_TYPE_CHECK,                .text = "eq",                            .value = COMPARISON_TYPE_EQUAL},
```

Since these are normal array values and not enums, I don't think you need to use the `[0] =` here. Same applies to all below cases.

UberWaffe replied:
Done.

crudelios reviewed (changes requested):
Yet another review batch.

Only one file left to review after this is done.

UberWaffe commented on src/scenario/scenario_event.c:
```diff
@@ -0,0 +1,208 @@
+        return 0;
+    }
+    
+    for (int i = 0; i < event->conditions.size; i++) {
```
Done.

crudelios replied:
I think you missed this one :wink:

UberWaffe commented on src/scenario/scenario_event.c:
```diff
@@ -0,0 +1,208 @@
+
+int scenario_event_conditional_execute(scenario_event_t *event)
+{
+    if (scenario_event_all_conditions_met(event))
+    {
```
Done.

UberWaffe commented on src/scenario/scenario_event_data.h:
```diff
@@ -0,0 +1,115 @@
+
+typedef struct {
+    condition_types type;
+    int32_t parameter1;
```
Done.

UberWaffe commented on src/scenario/scenario_events_controller.c:
```diff
@@ -0,0 +1,253 @@
+        array_size,
+        struct_size);
+
+    for (int i = 0; i < scenario_events.size; i++) {
```
Done.

UberWaffe commented on src/scenario/scenario_events_controller.c:
```diff
@@ -0,0 +1,253 @@
+    return scenario_events.size;
+}
+
+static void scenario_events_info_save_state(buffer *buf)
```
Done.

UberWaffe commented on src/scenario/scenario_events_controller.c:
```diff
@@ -0,0 +1,253 @@
+        &array_size,
+        &struct_size);
+
+    int link_type = 0;
```
Done.

UberWaffe commented on src/scenario/scenario_events_controller.c:
```diff
@@ -0,0 +1,253 @@
+        array_size,
+        struct_size);
+
+    for (int i = 0; i < scenario_events.size; i++) {
```
Just want to make sure I understand the array_foreach define correctly.
I could still use it for the outer for loop, correct?
Since it is just the equivalent of the code it defines array_foreach as?

crudelios replied:
Yes, you can use it for the outer loop.

UberWaffe commented on src/scenario/scenario_events_controller.c:
```diff
@@ -0,0 +1,253 @@
+
+void scenario_events_process_all(void)
+{
+    for (int i = 0; i < scenario_events.size; i++) {
```
Done.

UberWaffe commented on src/window/editor/scenario_condition_edit.c:
```diff
@@ -0,0 +1,170 @@
+    graphics_reset_dialog();
+}
+
+static void on_scroll(void)
```
No. Deleted.

UberWaffe commented on src/window/editor/scenario_event_details.c:
```diff
@@ -0,0 +1,309 @@
+    populate_list(0);
+}
+
+static void populate_list(int offset)
```
Done.

UberWaffe commented on src/window/editor/scenario_event_details.c:
```diff
@@ -0,0 +1,309 @@
+                button_border_draw(BUTTON_LEFT_PADDING, y_offset, BUTTON_WIDTH, DETAILS_ROW_HEIGHT, 1);
+            }
+
+            int font_color = COLOR_RED;
```
Done.

UberWaffe commented on src/window/editor/scenario_event_details.c:
```diff
@@ -0,0 +1,309 @@
+};
+
+static generic_button buttons[] = {
+    {SHORT_BUTTON_LEFT_PADDING, EVENT_REPEAT_Y_OFFSET, SHORT_BUTTON_WIDTH, 14, button_amount, button_none, 11, 0},
```
Done.

crudelios commented on src/scenario/condition_types/condition_types.c:
```diff
@@ -0,0 +1,282 @@
+    int max_months = condition->parameter3;
+    if (max_months < min_months) {
+        max_months = min_months;
+        condition->parameter3 = max_months;
```
Why are you changing `condition->parameter3` though? You should just swap `max_months` and `min_months` if they're reversed.

crudelios commented on src/scenario/scenario_events_controller.c:
```diff
@@ -0,0 +1,263 @@
+    }
+
+    if (repeat_max < repeat_min) {
+        log_error("Event maximum repeat is less than its minimum. Swapping the two values.", 0, 0);
```
```suggestion
        log_info("Event maximum repeat is less than its minimum. Swapping the two values.", 0, 0);
```

UberWaffe replied:
I was clearly not paying close attention at this point. Apologies.
Fixed.

crudelios commented on src/scenario/scenario_events_parameter_data.c:
Good job!

crudelios commented on src/scenario/scenario_events_xml.c:
```diff
@@ -0,0 +1,530 @@
+#include "scenario/scenario_events_controller.h"
+#include "scenario/scenario_events_parameter_data.h"
+#include "window/plain_message_dialog.h"
+#include <math.h>
```
Space between internal and external defines, please.

UberWaffe replied:
Done.

crudelios commented on src/scenario/scenario_events_xml.c:
```diff
@@ -0,0 +1,530 @@
+    }
+
+    int min = xml_parser_get_attribute_int("repeat_months_min");
+    if (!min) { min = 0; }
```
Please no same-line `if` statements. Same below for `max` and `max_repeats`.

crudelios commented on src/scenario/scenario_events_xml.c:
```diff
@@ -0,0 +1,530 @@
+    action_types type = get_action_type_from_element_name(name);
+    if (type == ACTION_TYPE_UNDEFINED) {
+        xml_import_log_error("Invalid action type specified");
+        log_info("Invalid action type specified", 0, 0);
```
Aren't you duplicating the log text here?

crudelios commented on src/scenario/scenario_events_xml.c:
```diff
@@ -0,0 +1,530 @@
+        data.error_line_number_text);
+}
+
+#pragma region ATTRIBUTE_PARSING
```
I understand why this may be useful, but it throws a warning on older (< 13) versions of GCC.

So it's best removed (both this and `endregion`).

UberWaffe replied:
Was more useful before I had refactored the functions down to about a third of the initial length. (Don't ask.)
Removed.

crudelios reviewed (changes requested):
Finally done!

Just some very small comments, then it's ready to merge.

UberWaffe commented:
Allright, last comments done.
I've squashed and rebased the branch.
Did a quick test of importing and ran the test scenario. Save and load on savegame and scenario.
All still appears in order.

crudelios commented:
You can check github actions for other builds, that's what I did.

Anyway, congratulations on finishing this! Time to merge 👏
